### PR TITLE
BI-1665 Store observations when importing new environments

### PR DESCRIPTION
**Pull request by davedrp**
# Description
[BI-1665](https://breedinginsight.atlassian.net/browse/BI-1665)

# Dependencies
sgn:  feture/BI-1665 branch
bi-web: develop branch

# Set Up
in your _sgn_local.conf_ add the key-value pair:
`brapi_observations_require_login 0`

# Testing

1. Import an Experiments & Observation file with at least one phenotype column with the corresponding timestamp column(s).
2. Confirm the import.
3. Insure that the data has been saved in he backing BrAPI server  

# Checklist:

- [x] I have performed a self-review of my own code
- [X] I have tested my code and ensured it meets the acceptance criteria of the story
- [X] I have tested that my code works with both the brapi-java-server and BreedBase
- [ ] I have create/modified unit tests to cover this change
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to documentation
- [ ] I have run TAF: _\<please include a link to TAF run>_

[BI-1665]: https://breedinginsight.atlassian.net/browse/BI-1665?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by nickpalladino — APPROVED** [review]

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java [hunk 1]
```diff
@@ -990,6 +1079,20 @@ private boolean validCategory(List<BrAPIScaleValidValuesCategories> categories,
      */
     private String yearsToSeasonDbId(List<String> years, UUID programId) {
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java:1080]: I know this is an existing method, but seems the name and parameters may be misleading, in that you pass in a list of years, but only get back a single seasonDbId.  Maybe this method should just be removed?

**Review by timparsons — CHANGES_REQUESTED** [review]
Some change requests.  Testing passed, but I didn't go into too many corner cases.

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java [hunk 2]
```diff
@@ -687,12 +692,98 @@ public void postBrapiData(Map<Integer, PendingImport> mappedBrAPIImport, Program
...
+                .filter(Objects::nonNull)
+                .distinct()
+                .map(PendingImportObject::getBrAPIObject)
+                .forEach(obsUnit -> updateObservationDbIds(obsUnit, programKey));
+
+        // Update ObservationVariable DbIds
+        List<Trait> traits = getTraitList(program);
+        for(PendingImportObject<BrAPIObservation> observation: this.observationByHash.values()){
+            traitsLoop:
+            for(Trait trait: traits){
```
[the remark this answers is not in the document]
Comment (davedrp) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java]: Thanks.  This has been updated

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java [hunk 3]
```diff
@@ -687,12 +692,98 @@ public void postBrapiData(Map<Integer, PendingImport> mappedBrAPIImport, Program
...
+            log.error(e.getMessage(), e);
+            throw new InternalServerException(e.toString(), e);
+        }
+        return traits;
+    }
 
+    private void updateObservationGermplasmDbId(BrAPIGermplasm germplasm) {
+        this.observationByHash.values().stream()
+                .filter(obs -> obs.getBrAPIObject().getGermplasmName() != null &&
+                        obs.getBrAPIObject().getGermplasmName().equals(germplasm.getGermplasmName()))
```
[the remark this answers is not in the document]
Comment (davedrp) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java]: Move logic to update the observation's gemplasmDbId to the updateObservationDbIds() method (see commit fb0b09a3d65a4bd0b85c9c502a5664d8192127de)

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java (hunk 1 above)
Comment (davedrp) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java]: Done.  (see comment 661aea0b412f535e13f5112b674a719365f70b50)

**Review by timparsons — APPROVED** [review]

**Pull request merged**